strict: function should not warn functions which have a non-simple parameter

**What version of ESLint are you using?**

- 2.12.0

**What parser (default, Babel-ESLint, etc.) are you using?**

- default

**Please show your full configuration:**

- nothing

**What did you do? Please include the actual source code causing the issue.**

```js
function foo(a = 0) {
}
```

    > eslint test.js --rule "strict:[error, function]" --env es6 --no-eslintrc --no-ignore

**What did you expect to happen?**

No error.
Because it's a syntax error in ES7/later if `"use strict"` is written there.

**What actually happened? Please include the actual, raw output from ESLint.**

```
test.js
  1:1  error  Use the function form of 'use strict'  strict

✖ 1 problem (1 error, 0 warnings)
```

--- a/docs/rules/strict.md
+++ b/docs/rules/strict.md
@@ -38,8 +38,6 @@ In the **CommonJS** module system, a hidden function wraps each module and limit
 
 In **ECMAScript** modules, which always have strict mode semantics, the directives are unnecessary.
 
-Since **ECMAScript 2016**, `"use strict"` directive in a function with non-simple parameter list is a syntax error. This rule disallows such `"use strict"` directives. See also the examples of [function option](#function)
-
 ## Rule Details
 
 This rule requires or disallows strict mode directives.
@@ -48,6 +46,8 @@ This rule disallows strict mode directives, no matter which option is specified,
 
 * `"sourceType": "module"` that is, files are **ECMAScript** modules
 * `"impliedStrict": true` property in the `ecmaFeatures` object
+
+This rule disallows strict mode directives, no matter which option is specified, in functions with non-simple parameter lists (for example, parameter lists with default parameter values) because that is a syntax error in **ECMAScript 2016** and later. See the examples of the [function](#function) option.
 
 ## Options
 
@@ -129,10 +129,6 @@ function foo() {
 
 function foo() {
 }
-```
-
-```js
-/*eslint strict: ["error", "function"]*/
 
 (function() {
     function bar() {

--- a/docs/rules/strict.md
+++ b/docs/rules/strict.md
@@ -38,6 +38,8 @@ In the **CommonJS** module system, a hidden function wraps each module and limit
 
 In **ECMAScript** modules, which always have strict mode semantics, the directives are unnecessary.
 
+Since **ECMAScript 2016**, `"use strict"` directive in a function with non-simple parameter list is a syntax error. This rule disallows such `"use strict"` directives. See also the examples of [function option](#function)
+
 ## Rule Details
 
 This rule requires or disallows strict mode directives.
@@ -139,6 +141,22 @@ function foo() {
 }());
 ```
 
+```js
+/*eslint strict: ["error", "function"]*/
+/*eslint-env es6*/
+
+// Illegal "use strict" directive in function with non-simple parameter list.
+// This is a syntax error since ES2016.
+function foo(a = 1) {
+    "use strict";
+}
+
+// We cannot write "use strict" directive in this function.
+// So we have to wrap this function with a function with "use strict" directive.
+function foo(a = 1) {
+}
+```
+
 Examples of **correct** code for this rule with the `"function"` option:
 
 ```js
@@ -150,8 +168,19 @@ function foo() {
 
 (function() {
     "use strict";
+
     function bar() {
     }
+
+    function baz(a = 1) {
+    }
+}());
+
+var foo = (function() {
+    "use strict";
+
+    return function foo(a = 1) {
+    };
 }());
 ```
 

--- a/lib/rules/strict.js
+++ b/lib/rules/strict.js
@@ -23,7 +23,9 @@ var messages = {
     unnecessary: "Unnecessary 'use strict' directive.",
     module: "'use strict' is unnecessary inside of modules.",
     implied: "'use strict' is unnecessary when implied strict mode is enabled.",
-    unnecessaryInClasses: "'use strict' is unnecessary inside of classes."
+    unnecessaryInClasses: "'use strict' is unnecessary inside of classes.",
+    nonSimpleParameterList: "'use strict' directive inside a function with non-simple parameter list throws a syntax error since ES2016.",
+    wrap: "Wrap this function in a function with 'use strict' directive."
 };
 
 /**
@@ -51,6 +53,26 @@ function getUseStrictDirectives(statements) {
     }
 
     return directives;
+}
+
+/**
+ * Checks whether a given parameter is a simple parameter.
+ *
+ * @param {ASTNode} node - A pattern node to check.
+ * @returns {boolean} `true` if the node is an Identifier node.
+ */
+function isSimpleParameter(node) {
+    return node.type === "Identifier";
+}
+
+/**
+ * Checks whether a given parameter list is a simple parameter list.
+ *
+ * @param {ASTNode[]} params - A parameter list to check.
+ * @returns {boolean} `true` if the every parameter is an Identifier node.
+ */
+function isSimpleParameterList(params) {
+    return params.every(isSimpleParameter);
 }
 
 //------------------------------------------------------------------------------
@@ -136,7 +158,9 @@ module.exports = {
                 isStrict = useStrictDirectives.length > 0;
 
             if (isStrict) {
-                if (isParentStrict) {
+                if (!isSimpleParameterList(node.params)) {
+                    context.report(useStrictDirectives[0], messages.nonSimpleParameterList);
+                } else if (isParentStrict) {
                     context.report(useStrictDirectives[0], messages.unnecessary);
                 } else if (isInClass) {
                     context.report(useStrictDirectives[0], messages.unnecessaryInClasses);
@@ -144,7 +168,11 @@ module.exports = {
 
                 reportAllExceptFirst(useStrictDirectives, messages.multiple);
             } else if (isParentGlobal) {
-                context.report(node, messages.function);
+                if (isSimpleParameterList(node.params)) {
+                    context.report(node, messages.function);
+                } else {
+                    context.report(node, messages.wrap);
+                }
             }
 
             scopes.push(isParentStrict || isStrict);
@@ -172,8 +200,13 @@ module.exports = {
 
             if (mode === "function") {
                 enterFunctionInFunctionMode(node, useStrictDirectives);
-            } else {
-                reportAll(useStrictDirectives, messages[mode]);
+            } else if (useStrictDirectives.length > 0) {
+                if (isSimpleParameterList(node.params)) {
+                    reportAll(useStrictDirectives, messages[mode]);
+                } else {
+                    context.report(useStrictDirectives[0], messages.nonSimpleParameterList);
+                    reportAllExceptFirst(useStrictDirectives, messages.multiple);
+                }
             }
         }
 

--- a/tests/lib/rules/strict.js
+++ b/tests/lib/rules/strict.js
@@ -67,6 +67,11 @@ ruleTester.run("strict", rule, {
             parserOptions: { ecmaVersion: 6 },
             options: ["function"]
         },
+        {
+            code: "(function() { 'use strict'; function foo(a = 0) { } }())",
+            parserOptions: {ecmaVersion: 6},
+            options: ["function"]
+        },
 
 
         // "safe" mode corresponds to "global" if ecmaFeatures.globalReturn is true, otherwise "function"
@@ -433,6 +438,80 @@ ruleTester.run("strict", rule, {
                 { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" },
                 { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" }
             ]
+        },
+
+        // Reports deprecated syntax: https://github.com/eslint/eslint/issues/6405
+        {
+            code: "function foo(a = 0) { 'use strict' }",
+            parserOptions: {ecmaVersion: 6},
+            options: [],
+            errors: ["'use strict' directive inside a function with non-simple parameter list throws a syntax error since ES2016."]
+        },
+        {
+            code: "(function() { 'use strict'; function foo(a = 0) { 'use strict' } }())",
+            parserOptions: {ecmaVersion: 6},
+            options: [],
+            errors: ["'use strict' directive inside a function with non-simple parameter list throws a syntax error since ES2016."]
+        },
+        {
+            code: "function foo(a = 0) { 'use strict' }",
+            parserOptions: {ecmaVersion: 6, ecmaFeatures: {globalReturn: true}},
+            options: [],
+            errors: [
+                "Use the global form of 'use strict'.",
+                "'use strict' directive inside a function with non-simple parameter list throws a syntax error since ES2016."
+            ]
+        },
+        {
+            code: "'use strict'; function foo(a = 0) { 'use strict' }",
+            parserOptions: {ecmaVersion: 6, ecmaFeatures: {globalReturn: true}},
+            options: [],
+            errors: ["'use strict' directive inside a function with non-simple parameter list throws a syntax error since ES2016."]
+        },
+        {
+            code: "function foo(a = 0) { 'use strict' }",
+            parserOptions: {ecmaVersion: 6},
+            options: ["never"],
+            errors: ["'use strict' directive inside a function with non-simple parameter list throws a syntax error since ES2016."]
+        },
+        {
+            code: "function foo(a = 0) { 'use strict' }",
+            parserOptions: {ecmaVersion: 6},
+            options: ["global"],
+            errors: [
+                "Use the global form of 'use strict'.",
+                "'use strict' directive inside a function with non-simple parameter list throws a syntax error since ES2016."
+            ]
+        },
+        {
+            code: "'use strict'; function foo(a = 0) { 'use strict' }",
+            parserOptions: {ecmaVersion: 6},
+            options: ["global"],
+            errors: ["'use strict' directive inside a function with non-simple parameter list throws a syntax error since ES2016."]
+        },
+        {
+            code: "function foo(a = 0) { 'use strict' }",
+            parserOptions: {ecmaVersion: 6},
+            options: ["function"],
+            errors: ["'use strict' directive inside a function with non-simple parameter list throws a syntax error since ES2016."]
+        },
+        {
+            code: "(function() { 'use strict'; function foo(a = 0) { 'use strict' } }())",
+            parserOptions: {ecmaVersion: 6},
+            options: ["function"],
+            errors: ["'use strict' directive inside a function with non-simple parameter list throws a syntax error since ES2016."]
+        },
+        {
+            code: "function foo(a = 0) { }",
+            parserOptions: {ecmaVersion: 6},
+            options: ["function"],
+            errors: ["Wrap this function in a function with 'use strict' directive."]
+        },
+        {
+            code: "(function() { function foo(a = 0) { } }())",
+            parserOptions: {ecmaVersion: 6},
+            options: ["function"],
+            errors: ["Use the function form of 'use strict'."]
         }
 
     ]

